crates/streamer/src/streamer/mod.rs

```diff
@@ -59,30 +62,51 @@ pub fn webm_stream(
 
     let cut_block_position = webm_itr.last_tag_position();
 
-    let framed = Framed::new(output_stream, ProtocolCodeC);
+    let ws_frame = Framed::new(output_stream, ProtocolCodeC);
 
     // ChannelWriter is a writer that writes to a channel
-    let (writer, receiver) = ChannelWriter::new();
-    let (error_sender, error_receiver) = tokio::sync::mpsc::channel(1);
+    let (chunk_writer, chunk_receiver) = ChannelWriter::new();
+    let (error_sender, error_receiver) = mpsc::channel(1);
+    let stop_notifier = Arc::new(Notify::new());
     spawn_sending_task(
-        framed,
-        receiver,
+        ws_frame,
+        chunk_receiver,
         match encode_writer_config.codec {
             cadeau::xmf::vpx::VpxCodec::VP8 => Some("vp8".to_owned()),
             cadeau::xmf::vpx::VpxCodec::VP9 => Some("vp9".to_owned()),
         },
         shutdown_signal,
         error_receiver,
+        Arc::clone(&stop_notifier),
     );
 
-    let mut header_writer = HeaderWriter::new(writer);
+    let mut header_writer = HeaderWriter::new(chunk_writer);
     debug!(?headers);
     for header in &headers {
         header_writer.write(header)?;
     }
 
     let mut encode_writer = header_writer.into_encoded_writer(encode_writer_config)?;
 
+    fn when_eof(
+        when_new_chunk_appended: &impl Fn() -> tokio::sync::oneshot::Receiver<()>,
+        stop_notifier: Arc<Notify>,
+    ) -> Result<WhenEofControlFlow, RecvError> {
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        let when_new_chunk_appended_receiver = when_new_chunk_appended();
+        tokio::spawn(async move {
+            tokio::select! {
+                _ = when_new_chunk_appended_receiver => {
+                    let _ = tx.send(WhenEofControlFlow::Continue);
+                },
+                _ = stop_notifier.notified() => {
+                    let _ = tx.send(WhenEofControlFlow::Break);
+                }
+            }
+        });
+        rx.blocking_recv()
+    }
```

Comment:
style: Put the function at the end.
https://github.com/Devolutions/IronRDP/blob/054f812f2e81d45737695fc3ee20f1465a31bd40/STYLE.md#helper-functions

Reply:
updated